typegate/src/planner/args.ts

```diff
@@ -899,11 +996,17 @@ class JsonSchemaBuilder {
   ): JSONSchema {
     const variantTypeIndices =
       node[variantsKey as keyof typeof node] as number[];
-    const variantSchemas = variantTypeIndices.map((typeIdx) =>
-      this.build(this.types[typeIdx])
-    );
+
+    const variantSchemas = variantTypeIndices.map((typeIdx) => {
+      const currentNode = this.types[typeIdx];
+      return visitedObject.has(currentNode.title)
+        ? refOf(currentNode)
+        : this.build(currentNode);
+    });
+    // Ref:
+    // https://json-schema.org/understanding-json-schema/reference/type.html
     return {
-      // ...trimType(node),
+      type: this.listUnionEitherTypes(node, variantsKey),
```

Comment:
I think `type` is not necessary for `anyOf` and `oneOf`. Is there a specific reason for this?

Reply:
yes, when the root `anyOf`/`oneOf` is optional and contain nested `anyOf`/`oneOf`, the parent `anyOf`/`oneOf` must be aware of the types in the next `anyOf`/`oneOf` and so on. In simple terms, we cannot do `someAnyOfOneOf.type = ["null"]` or `someAnyOfOneOf.type = ["object", "null"]`, we have to enumerate all types before the last `null`.